sable_history/src/pg_history_service.rs

```diff
@@ -0,0 +1,278 @@
+use tokio::sync::Mutex;
+use uuid::Uuid;
+
+use sable_network::prelude::*;
+
+use crate::schema::{channels, historic_users, messages};
+
+/// Implementation of [`HistoryService`] backed PostgreSQL
+pub struct PgHistoryService<'a> {
+    database_connection: &'a Mutex<AsyncPgConnection>,
+}
+
+impl<'a> PgHistoryService<'a> {
+    pub fn new(database_connection: &'a Mutex<AsyncPgConnection>) -> Self {
+        Self {
+            database_connection,
+        }
+    }
+}
+
+impl<'a> HistoryService for PgHistoryService<'a> {
+    async fn list_targets(
+        &self,
+        _user: UserId,
+        _after_ts: Option<i64>,
+        _before_ts: Option<i64>,
+        _limit: Option<usize>,
+    ) -> HashMap<TargetId, i64> {
+        // TODO: access control
+        // TODO: after_ts, before_ts, limit
+        match channels::dsl::channels
+            .select((
+                channels::dsl::id,
+                sql::<diesel::pg::sql_types::Uuid>(
+                    "SELECT MAX(id) FROM messages WHERE target_channel=channels.id",
+                ),
+            ))
+            .load_stream(&mut *self.database_connection.lock().await)
+            .await
+        {
+            Err(e) => {
+                tracing::error!("Could not get history channels: {e}");
+                return HashMap::new();
+            }
+            Ok(rows) => {
+                rows.map(|row| row.expect("Could not deserialize row"))
+                    .map(
+                        |(channel_id, max_message_id): (i64, Uuid)| -> (TargetId, i64) {
+                            let (seconds, _) = max_message_id
+                                .get_timestamp()
+                                .expect("messages.id is not a UUID7")
+                                .to_unix();
+                            (
+                                TargetId::Channel(ChannelId::from(Snowflake::from(
+                                    u64::try_from(channel_id).expect("channel id is negative"),
+                                ))),
+                                seconds
+                                    .try_into()
+                                    .expect("message's UNIX timestamp is negative"),
+                            )
+                        },
+                    )
+                    .collect()
+                    .await
+            }
+        }
+    }
+
+    async fn get_entries(
+        &self,
+        _user: UserId,
+        target: TargetId,
+        request: HistoryRequest,
+    ) -> Result<impl IntoIterator<Item = HistoricalEvent>, HistoryError> {
+        // TODO: access control
+        let TargetId::Channel(channel_id) = target else {
+            // TODO: PMs
+            return Err(HistoryError::InvalidTarget(target));
+        };
+
+        let mut connection_lock = self.database_connection.lock().await;
+
+        let db_channel_id = i64::try_from(channel_id.as_u64()).expect("channel id overflows u64");
```

Comment:
I've been preferring `as` conversions for this (to be replaced eventually with cast_signed/cast_unsigned), to just store and retrieve the bit pattern on the basis that we don't need the database server to be ordering IDs, just checking equality. 

Unless we do need the database to order them, in which case we need to change the existing code (and should probably enforce on creation that an id fits in 63 bits). Either way I think it's better to be consistent.

Reply:
ok for now, but in the future we should do it explicitly, eg. with bytemuck. Behavior of `as` depends on compilation profile, so it shouldn't be relied upon.